### ISSUE TITLE
docker ps inconsistency: master vs node

### ISSUE THREAD

**Issue by Garreat** [docker/classicswarm#2108]
Greetings Swarm community!
This issue seems somewhat related to:
[https://github.com/shipyard/shipyard/issues/728](https://github.com/shipyard/shipyard/issues/728)
[https://github.com/docker/swarm/issues/1468](https://github.com/docker/swarm/issues/1468)

**Output of `docker version`:**

```
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 16:16:33 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 16:16:33 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 4
Server Version: 1.10.2
Storage Driver: devicemapper
 Pool Name: docker-253:3-17700697-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.601 GB
 Data Space Total: 107.4 GB
 Data Space Available: 22.19 GB
 Metadata Space Used: 3.092 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.144 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-12-01)
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: overlay bridge null host
Kernel Version: 3.10.0-327.10.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 7.797 GiB
Name: n1
ID: WP23:YQON:7YTL:WFXI:WFWM:JL65:7GOA:FUKP:PISL:3IQ3:S3AH:VPBZ
WARNING: bridge-nf-call-ip6tables is disabled
Labels:
 role=node
Cluster store: consul://127.0.0.1:8500
Cluster advertise: x.x.x.x:2375
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
In my environment, all Swarm members are VSphere VMs within a single VLAN.

**Steps to reproduce the issue:**
1. Assuming you already have a Swarm environment...
2. Start a container.
3. Gracefully stop the container from inside. Don't use docker commands, simply lead container's PID 1 to a natural end.

**Describe the results you received:**
Swarm manager shows the container still running (incorrect):

```
dockkv1~# docker -H :3375 ps
(...)
485a2805a838        env-test        "/init /usr/bin/.."   13 hours ago        Up 13 hours        dockn1/test8
```

However, the node shows the container exited (correct):

```
dockn1~# docker ps -a
(...)
485a2805a838        env-test        "/init /usr/bin/.."   13 hours ago        Exited (0) 13 hours ago        test8
```

**Describe the results you expected:**
Swarm manager should correctly report the container state.
With this issue, it's not hard to imagine an orchestration scenario failing.

**Additional information you deem important (e.g. issue happens only occasionally):**
I use s6-overlay in the test container ([https://github.com/just-containers/s6-overlay](https://github.com/just-containers/s6-overlay)). BUT I have just reproduced this scenario with nginx container, using `nsenter` to go inside and then `nginx -s quit`. Result is the same: exited on the node, running on the master.
The containers are definitely stopped:

```
dockkv1~# docker -H :3375 stats ngx0 test8
CONTAINER           CPU %               MEM USAGE / LIMIT   MEM %               NET I/O             BLOCK I/O
ngx0                0.00%               0 B / 0 B           0.00%               0 B / 0 B           0 B / 0 B
test8               0.00%               0 B / 0 B           0.00%               0 B / 0 B           0 B / 0 B
```

The above containers left their overlay networks gracefully.
Simple `rm` command (without `-f`) is enough to fix this - despite containers being shown running.
The situation persists until an administrative action is taken manually.

**Comment by ezrasilvera** [comment]
I think I see where the issue may be coming from - Just to verify what do you get when you do "inspect" from Swarm on that exited container ?  Do you see the correct status ?

**Comment by Garreat** [comment]
Thanks for your reply

```
dockkv1~# docker -H :3375 inspect ngx0
[
    {
        "Id": "0091b6c66b56fe573f1552aa10e0b1585b7539a8fb721d2dc53f4cf05be5b31a",
        "Created": "2016-04-11T13:23:33.920712225Z",
        "Path": "nginx",
        "Args": [
            "-g",
            "daemon off;"
        ],
        "State": {
            "Status": "exited",
            "Running": false,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 0,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2016-04-11T13:23:35.033143466Z",
            "FinishedAt": "2016-04-11T13:24:47.98456338Z"
        },
```
It looks correct.

**Comment by dongluochen** [comment]
What is swarm version? Are there logs from Swarm like "Error monitoring events"? In Swarm container state update is event driven. When a container's state change, Swarm should receive an event from the corresponding node and refresh the container to retrieve the container state. There was issue on losing event stream from nodes. It should be fixed in Swarm 1.1.3.

**Comment by Garreat** [comment]
Thanks for your reply
```
dockkv1~# docker -H :3375 info
Containers: 16
 Running: 16
 Paused: 0
 Stopped: 0
Images: 17
Server Version: swarm/1.1.3
(...)
```
I have debug flag among daemon parameters: `ExecStart=/usr/bin/docker daemon -D `.
I have ran `dockkv1~# journalctl -u docker > debug.log; cat debug.log | grep -i error`
and received no output.
The file `debug.log` is about 20MB large, containing almost entirely this:
```
Apr 11 23:00:38 dockkv1 docker[5611]: time="2016-04-11T23:00:38.001867182+02:00" level=debug msg="Watch triggered with 5 nodes" discovery=consul
Apr 11 23:00:38 dockkv1 docker[5611]: time="2016-04-11T23:00:38.792088792+02:00" level=debug msg="Watch triggered with 5 nodes" discovery=consul
Apr 11 23:00:45 dockkv1 docker[5611]: time="2016-04-11T23:00:45.185001915+02:00" level=debug msg="Watch triggered with 5 nodes" discovery=consul
```

**Comment by dongluochen** [comment]
How about Swarm logs? You can also enable swarm debugging with `swarm -l debug manage ...`. Are there errors?

**Comment by ezrasilvera** [comment]
@dongluochen The problem is that in order to help with performance the periodic refresh loop uses RefreshContainers(false) --> which means there will be no inspect and no full update in case the container exists (which in this case it does).
This means that in general in case we loose the event (we do full update on event) the refresh loop will not pick ANY change (state, info, etc..) ...

**Comment by dongluochen** [comment]
@ezrasilvera Yes this is a problem if manager misses events. I'm thinking to force a `full` update in the refreshloop on a predefined interval. WDYT?

**Comment by Garreat** [comment]
I have rebuilt the cluster using `swam -l debug manage` for swarm binary container.
Reproduced the issue as expected.
`dockkv2~# docker logs swarm | grep -i error` = no output

I'm not an expert on Docker internals but I think I understand your point @ezrasilvera .
@dongluochen maybe this is a kind of collateral damage from 'fastest orchestrator' race. I can't remember having similar issue across previous Docker versions (1.4+). I can see the considerations though,
Maybe a nice trade-off would be to provide a 'high-awareness' switch for Swarm master?
My precious Swarm isn't going to scale into tens thousands containers.

**Comment by ezrasilvera** [comment]
I wouldn't recommend it  Remember that this will imply doing inspect on ALL containers on each loop this is a huge performance hit and what is worse is that it will introduce linear degradation in performance as the number of containers grow.
Maybe we can just add a status change check after calling ListContainers and force a "full" update if there is a state change ...
I can try it out later on  and push a PR....

**Comment by dongluochen** [comment]
`inspect on ALL containers on each loop`, no we don't want to do that. It'd be a much larger interval, like 5 minutes. Recently we add this to refresh Engine spec because some Engine spec changes do not trigger events.

**Comment by Garreat** [comment]
For now my workaround is to `docker rm` before each `docker run` (only 'fake running' would actually get deleted). Could also incorporate some `docker start` but that's kinda stateful experience.

Anyway, I believe we can call this a bug... the current output of `ps` on Swarm master is just wrong (misleading).

**Comment by doronp** [comment]
This issue is about (eventual) consistency if I get it right. 

Either one acknowledges that the state represented by Swarm can be stale for some shot amount of time (relatively to a container state change on an engine in the cluster) or one acknowledges that once he send a command to Swarm he should wait (for full update). 

Maybe the solution is to let the user decide if he can live with eventual consistency or not.
So a combination or configurable flag of when to perform a full update (as @dongluochen suggests) may make most of the users happy. 

I can help with the flag only or with the whole issue if you like.

**Comment by dongluochen** [comment]
> I have rebuilt the cluster using swarm -l debug switch for swarm binary containers.
Reproduced the issue as expected.

@Garreat Can you provide the steps to reproduce the problem? How many nodes in the swarm cluster? What commands do you use to get to this state?

**Comment by dongluochen** [comment]
@ezrasilvera @doronp I like your comments. Can you sync up on PR preparation? 

IMHO we might have 2 problems here. The first problem is container state update. Swarm manager subscribes to Engine event stream to refresh corresponding container state. It should keep container state fresh. We should fix any problem from this. 

The second problem is event misses on manager or for some reason Engine doesn't trigger the corresponding event. Container state would be wrong. Adding a `sync` with some interval to correct them makes sense. It shouldn't take much resource to do it.

**Comment by ezrasilvera** [comment]
@dongluochen I'm not sure that (1) is relevant to the discussion. If indeed there is a bug in how we handle events we should obviously fix as we do with every other issue :-)
We want to see if we identify a scenario that can lead to out-of sync info/state. If we believe that events can be missed or that the engine may not generate events then indeed this is such a scenario, and we will get out of sync even if there is no bug in he event handling.
If as I indicated we can inspect the state during ListContainers then the problem can be easily solved using the existing refresh loop (I'm checking this). If this information is returned only through Inspect then as you suggested we can add a full refresh loop an a much longer period (minutes).
We can expose the time interval as environment variable so it will allow system to tune this interval (even add some learning and feedback if someone wants to be extremely cleaver :) )

**Comment by dongluochen** [comment]
> I'm not sure that (1) is relevant to the discussion. If indeed there is a bug in how we handle events we should obviously fix as we do with every other issue

Totally agree. That's why I ask for the reproduce steps. 

> We want to see if we identify a scenario that can lead to out-of sync info/state.

Yes it'd help how to resolve it.

**Comment by ezrasilvera** [comment]
Agree - I tried to do some exits and couldn't reproduce :(   Seems Swarm got the correct event on my tests :-)  I'm not an expert in go channels and can't comment on theoretical event lost ...
I guess the chances for that will increase as the number of engine increases

**Comment by doronp** [comment]
I noticed that:
https://github.com/docker/swarm/blob/master/cluster/engine.go#L769

(No refresh if update fails)

So @Garreat Can you grep your log for 
"Update engine specs failed"

And let us know ?

**Comment by ezrasilvera** [comment]
@doronp that's not relevant to the problem at hand
if the engine is not healthy the inspect he did wouldn't have worked (and judging by his report it was working as expected) 
If we believe events can get lost (and on large scale cluster this should  be our assumption), or events may not sent/generated correctly (on the engine side) then we must have some corrective sync actions and the only way to do it is either trigger or periodically activate sync.

**Comment by doronp** [comment]
> that's not relevant to the problem at hand
That's an easy test no need to make assumptions when one can test.

> If we believe events can get lost ...
We believe and I think we are all on agreement that we should add periodic sync and expose the refresh iteration interval for configuration. 

I'd even consider making full refresh loop on possibly another interval to be on the safe side:
https://github.com/docker/swarm/blob/master/cluster/engine.go#L781
(true on the refresh)

**Comment by ezrasilvera** [comment]
@doronp that was the proposal ..... do a full refresh loop at an interval defined by the user (see couple of comments  above)

**Comment by Garreat** [comment]
tldr: can't reproduce after restarting docker daemon on each Swarm member (managers and nodes)

infrastructure (vSphere VMs ):
- dockkv1, dockkv2; docker daemon running consul server and swarm manage
- dockgw1, dockn1, dockn2: docker daemon running consul client and swarm join

what I've done before:
- (every member) docker rm -f $(docker ps -aq)
- rebuild Consul infrastructure
- rebuild Swarm infrastructure
- created 10-12 containers (`nginx` and `test` container mentioned in first post; the `test` container takes 200-800MB depending on load)
- 'softly' stop any of those without using `docker` commands (from inside container)
- compare `ps` output: swarm manager vs node engine
- issue present for both `test` and `nginx` image containers

what I've done 30 minutes ago:
- (every member, manager or node) systemctl restart docker
- all above steps
- *issue gone*

It's worth mentioning that the daemons were running for about 2 weeks straight before the systemctl restart docker. A few `swarm join` restarts did occur (no sudden host/daemon restart though). Also, the system clock differs among member machines (1-5min differences). Sadly I can't tell if this time diff was there before. There is no NTP setup. For now I left it that way.

@ezrasilvera 
> if the engine is not healthy the inspect he did wouldn't have worked (and judging by his report it was working as expected)
Yes, (on swarm manager) the inspect of fake-running container has returned the 'exited' status.

Confusion... being unable to reproduce makes me feel stupid :/. Was trivial before... I believe there is something more into it. Please keep in mind the issues mentioned on topic start, this does happen:
[https://github.com/shipyard/shipyard/issues/728](https://github.com/shipyard/shipyard/issues/728)
[https://github.com/docker/swarm/issues/1468](https://github.com/docker/swarm/issues/1468)

I will provide the actual commands once my VPN is back up -.-' .... Those didn't really change (added only swarm _-l debug_ manage). I will also try some stuff to reproduce. Hopefully then I could find `"Update engine specs failed"` requested by @doronp? (of course the old log is gone -.-')
Tomorrow at the site I'll have more insight. Getting sleepy now.

In the meantime, thank you all for your bright input @ezrasilvera @dongluochen @doronp. It's awesome having community this developed :).

**Comment by ezrasilvera** [comment]
@Garreat that's 100% fine!  your post actually triggered a discussion on a more general problem so it was very helpful !!  :-)

**Comment by ezrasilvera** [comment]
@dongluochen @doronp and I synced up on the PR and we think that our combined approach will address the general issue  of sync between Swarm data and Engine/Container data
1) Full update loop with a user defined time interval
2) Update the Container fields (e.g., status, ports, etc..) returned during the ListContainer (used in refreshContainers) - currently if the container exists nio update is being done 
We will push this shortly  (might separate into two PRs  )..

**Comment by chrodriguez** [comment]
Hi everyone! I've recently installed a swarm cluster and evidenced some kind of the behavior described by this issue, but in a different way.

First I will describe my infraestructure:

1. Consul HA + Swarm manager HA
  1. SERVER 1: consul server + swam manager
  2. SERVER 2: consul server + swarm manager 
 3. SERVER 3: consul server + swarm manager  
2. Swarm nodes
  1. SERVER 1: swarm node
  2. SERVER 2: swarm node

Each swarm node runs **registrator** and every new docker container run inside a swarm manager were run with an `always restart policy`

I've started 6 nginx containers this way and reboot one of the swarm nodes server (where there were at least one container running), and then `docker -H :3376 ps ` will consider that rebooted server containers are not running, but `docker ps` in that swarm node reports container is running. Moreover, those containers shows a running status when they are inspected with `docker -H :3376 inspect xxxx`

I don't know how this issue may impact a production infrastructure, but is not trivial to read ps output when reports invalid information from the swarm cluster.

I hope my description helps solving this issue, thanks in advance!

**Comment by mixman** [comment]
I ran into the same issue today. A cluster of two Managers and three connected Nodes. All nodes reported as Healthy, yet the primary manager would not list the running containers of the Nodes while the Nodes had them running. Restarting docker on each Node made Swarm aware of the containers again.

Docker version 1.11.1, build 5604cbe
Docker Swarm version is swarm/1.2.0

Currently Swarm 1.2.0 is the latest version that works with private registries (Docker Hub).

**Comment by rreinurm** [comment]
I have same issue. My environment

**Output of docker version**
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64

**Output of Docker Info**
Containers: 17
 Running: 14
 Paused: 0
 Stopped: 3
Images: 477
Server Version: 1.11.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 662
 Dirperm1 Supported: false
Logging Driver: fluentd
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host overlay bridge
Kernel Version: 3.13.0-63-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.67 GiB
Name: d1
ID: 7RCJ:QT67:CVYC:JBTV:KCGW:CVO7:M4R2:IJPH:24FE:ZR3J:VMHO:TKE4
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Cluster store: consul://consul.service.consul:8500/swarm
Cluster advertise: x.x.x.x:2376

**Output from Swarm Version**
Server:
 Version:      swarm/1.2.2
 API version:  1.22
 Go version:   go1.5.4
 Git commit:   34e3da3
 Built:        Mon May  9 17:03:22 UTC 2016
 OS/Arch:      linux/amd64

**Out from Swarm Info**
docker info
Containers: 70
 Running: 57
 Paused: 0
 Stopped: 13
Images: 153
Server Version: swarm/1.2.2
Role: primary
Strategy: spread
Filters: health, port, containerslots, dependency, affinity, constraint
Nodes: x
 d1: d1.fq.dn:2376
  └ ID: 7RCJ:QT67:CVYC:JBTV:KCGW:CVO7:M4R2:IJPH:24FE:ZR3J:VMHO:TKE4
  └ Status: Healthy
  └ Containers: 17
  └ Reserved CPUs: x / x
  └ Reserved Memory:xx GiB / xx GiB
  └ Labels: executiondriver=, kernelversion=3.13.0-63-generic, operatingsystem=Ubuntu 14.04.4 LTS, storagedriver=aufs
  └ Error: (none)
  └ UpdatedAt: 2016-05-30T12:08:06Z
  └ ServerVersion: 1.11.1
Plugins:
 Volume:
 Network:
Kernel Version: 3.13.0-63-generic
Operating System: linux
Architecture: amd64
CPUs: xx
Total Memory: xx GiB
Name: d2
Docker Root Dir:
Debug mode (client): false
Debug mode (server): false
WARNING: No kernel memory limit support


After Docker Daemon is restarted ( including Swarm Agent which is running in container) all services are displayed with Exit status from Swarm. On the Docker host all containers have state running. If I restart all containers through Swarm, then status is displayed correctly.

**Comment by mboudet** [comment]
Same issue here when launching container on Swarm with Docker Compose

**Docker Swarm Version**
```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:38:58 2016
 OS/Arch:      linux/amd64

Server:
 Version:      swarm/1.2.3
 API version:  1.22
 Go version:   go1.5.4
 Git commit:   eaa53c7
 Built:        Fri May 27 17:25:03 UTC 2016
 OS/Arch:      linux/amd64

```
**Docker Info**
```
Containers: 3
 Running: 3
 Paused: 0
 Stopped: 0
Images: 4
Server Version: swarm/1.2.3
Role: primary
Strategy: spread
Filters: health, port, containerslots, dependency, affinity, constraint
Nodes: 1
 VM-4601: 192.168.2.61:2375
  └ ID: 5H66:YLHH:Z2JN:JRI4:7Q64:GPTH:ALX7:YV6G:PVB4:G4UQ:QMTX:VEGV
  └ Status: Healthy
  └ Containers: 3
  └ Reserved CPUs: 0 / 2
  └ Reserved Memory: 0 B / 2.061 GiB
  └ Labels: executiondriver=, kernelversion=3.16.0-4-amd64, operatingsystem=Debian GNU/Linux 8 (jessie), storagedriver=aufs
  └ UpdatedAt: 2016-06-06T08:39:47Z
  └ ServerVersion: 1.11.1
Plugins: 
 Volume: 
 Network: 
Kernel Version: 3.16.0-4-amd64
Operating System: linux
Architecture: amd64
CPUs: 2
Total Memory: 2.061 GiB
Name: 57b3cbd6c171
```
**Docker ps from Host**
```
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               
7df7257304eb       bioconda/abyss     "/bin/bash"         58 minutes ago      Up 52 minutes                           
eb513e73d85d       bioconda/bamhash   "/bin/bash"         About an hour ago   Up About an hour                        
```
**Docker ps -a from Node**
```
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                         PORTS               
7df7257304eb        bioconda/abyss     "/bin/bash"         About an hour ago   Exited (0) About an hour ago                         
eb513e73d85d        bioconda/bamhash   "/bin/bash"         About an hour ago   Exited (0) About an hour ago                       
```
After the Docker Daemon is restarted, the containers are correctly shown as stopped.

**Comment by nktl** [comment]
Same/very similar problem as @rreinurm, @mixman and @chrodriguez 

It looks like Swarm is not able to track state of the containers if they are manipulated (deployed, restarted, etc) outside of Swarm - i.e. by running operations on docker hosts directly.

**Comment by sukyology** [comment]
hey guys!

so what is happening with this issue? 

i have similar issue 

i have installed shipyard and manually deployed two custom containers. 

![docker_swarm_not_updating](https://cloud.githubusercontent.com/assets/6030180/16757361/713ba01e-4846-11e6-81f4-23ed5d0443af.png)


this is happening after i reboot the node. 
there are certain containers that get fully updated(including status) and certain containers do not get updated... 

i couldn't tell the difference between containers that get fully updated and ones that don't. 

just curious what kind of plan you guys have with this issue. 
are u guys going to do full refresh periodically? 
or does user have to do something to trigger full refresh for a container? 
or haven't you guys decide yet. 

please let me know. 
thanks.

**Comment by nktl** [comment]
Probably not much development is going to happen here, as Docker is now moving to native swarm support with swarmkit (from 1.12) - which should mitigate this problem (if designed/implemented properly)

**Comment by sukyology** [comment]
@nktl thanks body.

**Comment by antoinetran** [comment]
Dear all,
We have the same error with Swarm (classic mode 1.2.6) + docker-17.06-ce + 1 SwarmManager / 1 consul:

```
[ivvts@fl2p-s02 ~]$ sudo docker ps -a|grep vibra
54f9ad4130dd        fl2p-s01.ts-l2pf.cloud-omc.org:5000/thales/baseimage:0.2                                                                               "echo Sleeping 1s..."     13 minutes ago      Exited (0) 13 minutes ago                                  vibrant_nightingale
[ivvts@fl2p-s02 ~]$ docker -H [SwarmManager]:4000 ps -a|grep vibra
54f9ad4130dd        fl2p-s01.ts-l2pf.cloud-omc.org:5000/thales/baseimage:0.2                                                                               "echo Sleeping 1s..."     13 minutes ago      Up 13 minutes                                                                                                                                                                  fl2p-s02.ts-l2pf.cloud-omc.org/vibrant_nightingale
```
Unfortunately, we cannot migrate to Swarm mode because of major incompatibilities for now. Does someone work on this?

**Comment by nktl** [comment]
This won't be fixed, swarm is dead - get rid of it ASAP.
All container action happens in Kubernetes now.

**Comment by antoinetran** [comment]
@nktl that is a bit difficult for now, especially when Kubernetes integration in docker is still really recent (in production about 2 months ago), and not in Linux (yet?) for docker-ce (contrary to docker-ee). Moreover, docker didn't officially give up about Swarm ([source](https://www.docker.com/enterprise-edition))

> Docker Enterprise Edition allows you to run Swarm and Kubernetes interchangeably and provides a secure and simple way to operationalize Kubernetes.

**Comment by nktl** [comment]
The sad truth is that Docker dropped the ball big time with Swarm, failing to deliver feasible enterprise-grade solution on time (in fact, it still does not work well and lacks basic features)

This void was perfectly filled with Kubernetes (and its various flavors, like OpenShift). Kubernetes has been using Docker runtime in production for a long, long time.  Both docker-ce / docker-ee are irrelevant now, the container runtime is directly integrated into Kubernetes.

Additionally, there are projects ongoing (like CRI-O) that aim to decouple container runtime from other management aspects (buliding containers, etc) and weed out traditional Docker completely. This will be soon embraced naively by Kubernetes and make Docker even less relevant.
